src/models/forecasts.jl

```diff
@@ -4,48 +4,99 @@ const Forecasts = Vector{<:Forecast}
-#   deserialization.
-#
-# The code below converts the forecast data to a different format to enable deserialization.
-
-struct _SystemForecastArrayForJSON
-    resolution::Dates.Period
-    initialtime::Dates.DateTime
-    forecasts::Forecasts
-    forecast_types::Vector{String}  # Encode the exact type so that deserialization doesn't have
-                            # to infer what it is.
+
+struct _ForecastKey
+    initial_time::Dates.DateTime
+    forecast_type::DataType
 end
 
-struct _SystemForecastsForJSON
-    forecasts::Vector{_SystemForecastArrayForJSON}
+const _Forecasts = Dict{_ForecastKey, Vector{<:Forecast}}
+
+function _get_forecast_initial_times(data::_Forecasts)::Vector{Dates.DateTime}
+    initial_times = Set{Dates.DateTime}()
+    for key in keys(data)
+        push!(initial_times, key.initial_time)
+    end
+
+    return sort!(Vector{Dates.DateTime}(collect(initial_times)))
 end
 
-function JSON2.write(io::IO, forecasts::SystemForecasts)
-    return JSON2.write(io, encode_for_json(forecasts))
+function _add_forecast!(data::_Forecasts, forecast::T) where T <: Forecast
+    key = _ForecastKey(forecast.initialtime, T)
+    if !haskey(data, key)
+        data[key] = Vector{T}()
+    end
+
+    push!(data[key], forecast)
 end
 
-function JSON2.write(forecasts::SystemForecasts)
-    return JSON2.write(encode_for_json(forecasts))
+"""Container for forecasts and their metadata. Implementation detail that is not exported.
+Functions to access the data should go through the System."""
+struct SystemForecasts
+    data::_Forecasts
+    initial_time::Dates.DateTime
+    resolution::Dates.Period
+    horizon::Int64
+    interval::Dates.Period
 end
 
-function encode_for_json(system_forecasts::SystemForecasts)
-    forecasts_for_json = Vector{_SystemForecastArrayForJSON}()
-    for (issue_time, forecasts) in system_forecasts
-        forecast_types = [strip_module_names(typeof(x)) for x in forecasts]
-        push!(forecasts_for_json, _SystemForecastArrayForJSON(issue_time.resolution,
-                                                              issue_time.initialtime,
-                                                              forecasts,
-                                                              forecast_types))
+"""Constructs SystemForecasts from the previous version of the data structure."""
+function SystemForecasts(old_forecasts::SystemForecastsDeprecated)
+    forecasts = _Forecasts()
+    initial_time = Dates.DateTime(Dates.Second(1))
+    resolution = Dates.Period(Dates.Second(1))
+    initialized = false
+    horizon::Int64 = 0
+
+    for forecast in Iterators.flatten(values(old_forecasts))
+        if !initialized
+            initial_time = forecast.initialtime
+            resolution = forecast.resolution
+            horizon = length(forecast)
+            initialized = true
+        else
+            if forecast.resolution != resolution
+                # TODO parsing code currently produces multiple resolutions.
```

Comment:
I'll need help from you guys in order to proceed.  I can work on the parsing code if needed.

Reply:
Let's leave like this for now until the time series parsing is resolved.